cpp/src/arrow/ipc/memory.cc

```diff
@@ -144,5 +144,29 @@ Status MemoryMappedSource::Write(int64_t position, const uint8_t* data, int64_t
   return Status::OK();
 }
 
+MockMemorySource::MockMemorySource(int64_t size) : size_(size) {}
+
+Status MockMemorySource::Close() {
+  return Status::OK();
+}
+
+Status MockMemorySource::ReadAt(
+    int64_t position, int64_t nbytes, std::shared_ptr<Buffer>* out) {
+  return Status::OK();
+}
+
+Status MockMemorySource::Write(int64_t position, const uint8_t* data, int64_t nbytes) {
+  pos_ = position + nbytes;
```

Comment:
I think it is worth clarifying in the header, that pos_ is always a function of the last Write call.  A more complete (but possibly not necessary write now) would actually track all non-overlapping writes to to verify the actual number of bytes  written)